core/utils/thread_control.go

```diff
@@ -0,0 +1,74 @@
+package utils
+
+import (
+	"context"
+	"errors"
+	"sync"
+	"sync/atomic"
+)
+
+var (
+	_ ThreadControl = &threadControl{}
+
+	ErrThreadLimitReached = errors.New("thread limit reached")
+)
+
+// ThreadControl is a helper for managing a group of goroutines.
+type ThreadControl interface {
+	// Go starts a goroutine and tracks the lifetime of the goroutine.
+	Go(fn func(context.Context)) error
+	// Close cancels the context and waits for all tracked goroutines to exit.
+	Close()
+}
+
+func NewThreadControl(limit int) *threadControl {
+	tc := &threadControl{
+		stop:  make(chan struct{}),
+		limit: int32(limit),
+	}
+
+	return tc
+}
+
+type threadControl struct {
+	threadsWG sync.WaitGroup
+
+	limit   int32
+	running atomic.Int32
+
+	stop StopChan
+}
+
+func (tc *threadControl) Go(fn func(context.Context)) error {
+	if err := tc.add(); err != nil {
+		return err
+	}
+
+	go func() {
+		defer tc.done()
+		ctx, cancel := tc.stop.NewCtx()
+		defer cancel()
+		fn(ctx)
+	}()
+
+	return nil
+}
+
+func (tc *threadControl) Close() {
+	close(tc.stop)
+	tc.threadsWG.Wait()
+}
+
+func (tc *threadControl) add() error {
+	if tc.running.Add(1) > tc.limit {
+		tc.running.Add(-1)
+		return ErrThreadLimitReached
```

Comment:
Is this limit pattern used in the codebase already? It seems difficult to use, since a caller that is willing to wait just gets rejected and has to retry, with an incentive to spam very quick to get in under the limit. Do we have a real reason to worry about limits here? If so, a common pattern is a set of worker routines that pull from a queue. Then calls can just block until the work is done (or context expires).

Reply:
If limit is necessary, maybe consider using `https://pkg.go.dev/golang.org/x/sync/semaphore`, and exposing API for both blocking and non-blocking cases.

Reply:
I'll remove the limit as we don't really need it right now